### PR TITLE
Feature/free threaded python

### PR DESCRIPTION
**Pull request by sametd**
## Free-Threaded Python Support
This PR ises #31  as its base. This PR: Remove the GIL as a bottleneck for tensogram's Python bindings. Multiple Python threads can now encode, decode, scan, and validate in true parallel — on both GIL-enabled Python (via `py.detach()`) and free-threaded Python 3.13t/3.14t (no GIL at all).
### What changed
**Rust core** — all `TensogramFile` read methods changed from `&mut self` to `&self`. Local backend uses `OnceLock` for lock-free lazy scanning. Remote backend uses `Mutex<RemoteState>` with fine-grained locking so I/O runs without holding the lock. Only `append()` requires exclusive access. This benefits all consumers: Rust, Python, FFI, CLI.
**Python bindings** — upgraded to PyO3 0.28 with `#[pymodule(gil_used = false)]`. 24 `py.detach()` calls release the GIL during every Rust computation path. `PyBackedBytes` eliminates buffer copies on decode/scan/validate inputs. `GILOnceCell` replaced with `PyOnceLock` for thread-safe initialization.
**Tests** — 23 thread-safety tests covering concurrent encode/decode across codecs, shared-handle reads, scan, validate, file ops, streaming encoder, and read+append borrow checking.
**CI** — Python 3.13t job with `PYTHON_GIL=0` verification.
**Benchmarks** — threading scaling suite and a head-to-head comparison against ecCodes C on the standard weather data pipeline (10M float64, 24-bit packing + szip).
### Results
With packing + szip (the real-world weather pipeline), Tensogram from Python at 2 threads already exceeds ecCodes C single-threaded decode throughput. At 8 threads on free-threaded Python: 4.9x ecCodes decode, 2.7x ecCodes encode.
Validate scales to 9.7x at 16 threads. Decode-range scales to 8.6x. Small messages (16K elements) scale to 8.4x.

### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.
 

Docs Preview
https://sites.ecmwf.int/docs/dev-section/tensogram/pull-requests/PR-32

### CODE REVIEW

File: crates/tensogram-python/src/lib.rs [hunk 1]
```diff
@@ -854,12 +971,13 @@ impl PyBufferIter {
...
         self.index += 1;
-        let msg_bytes = &self.buf[offset..offset + length];
+        let msg_bytes = self.buf[offset..offset + length].to_vec();
         let options = DecodeOptions {
             verify_hash: self.verify_hash,
             ..Default::default()
         };
-        let (global_meta, data_objects) = decode(msg_bytes, &options).map_err(to_py_err)?;
+        let (global_meta, data_objects) =
+            py.detach(|| decode(&msg_bytes, &options).map_err(to_py_err))?;
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-python/src/lib.rs:980]: @copilot The copy is required. py.detach() demands Send on the closure, and &self.buf[offset..offset + length] borrows from &mut self on a #[pyclass], which is not Send. Removing .to_vec() would fail to compile. The per-message allocation is the correct trade-off for releasing the GIL during decode.

File: crates/tensogram-python/src/lib.rs [hunk 2]
```diff
@@ -20,6 +21,8 @@ use tensogram_core::{
     GlobalMetadata, HashAlgorithm, StreamingEncoder, TensogramError, TensogramFile, RESERVED_KEY,
 };
 
+type PyObject = Py<PyAny>;
+
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-python/src/lib.rs:25]: @copilot Not a bug. PyO3 0.28 removed PyObject from the prelude (use pyo3::prelude::* no longer includes it). The type alias is required — removing it would fail to compile.

File: tensogram-zarr/src/tensogram_zarr/store.py [hunk 3]
```diff
@@ -324,15 +360,32 @@ async def list_dir(self, prefix: str) -> AsyncIterator[str]:
...
             raise OSError(f"failed to open TGM file {self._path!r}: {exc}") from exc
 
-        with f:
+        if is_remote:
+            try:
+                self._scan_remote(f)
+            except Exception:
+                f.__exit__(None, None, None)
+                raise
+        else:
```
[the remark this answers is not in the document]
Comment (sametd) [reply tensogram-zarr/src/tensogram_zarr/store.py:388]: @copilot Intentionally not adding message_count() here. For remote files, message_count() triggers scan_all which scans every message boundary — O(N) HTTP requests for N messages. The remote scan path is designed to be lazy: file_decode_descriptors(message_index) only scans up to the requested message and raises on out-of-range, keeping the open cost O(1). Adding an eager count would regress the remote open from one HTTP request to hundreds for large files.

File: crates/tensogram-core/src/decode.rs [hunk 4]
```diff
@@ -128,14 +128,15 @@ pub fn decode_object(
...
+/// per range.  The descriptor is included so callers can determine
+/// the dtype without a separate lookup.
 pub fn decode_range(
     buf: &[u8],
     object_index: usize,
     ranges: &[(u64, u64)],
     options: &DecodeOptions,
-) -> Result<Vec<Vec<u8>>> {
+) -> Result<(DataObjectDescriptor, Vec<Vec<u8>>)> {
     let msg = framing::decode_message(buf)?;
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-core/src/decode.rs:140]: @copilot This change is in the feature/remote-polish base branch, not introduced by this PR. The return type change was intentional — decode_range callers (Python bindings, xarray backend, FFI) all need the descriptor to interpret the returned byte ranges. Will document in the changelog when feature/remote-polish is merged.

File: crates/tensogram-core/src/file.rs [hunk 5]
```diff
@@ -191,47 +285,118 @@ impl TensogramFile {
...
+        match &self.backend {
+            Backend::Local { path, .. } => Some(path),
+            #[cfg(feature = "remote")]
+            Backend::Remote(_) => None,
+        }
+    }
+
+    pub fn source(&self) -> String {
+        self.backend.source_string()
     }
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-core/src/file.rs:318]: @copilot Also from the feature/remote-polish base branch. The change to Option<&Path> is intentional — remote files don't have a local path. source() returns a string representation for both local and remote. Will document when that branch is merged.
